pkg/statedb2/txn.go

```diff
@@ -0,0 +1,369 @@
+}
+
+func (txn *txn) addDeleteTracker(meta TableMeta, trackerName string, dt deleteTracker) error {
+	if txn.rootReadTxn == nil {
+		return fmt.Errorf("transaction is closed")
+	}
+	table, ok := txn.tables[meta.Name()]
+	if !ok {
+		return fmt.Errorf("table %q not locked for writing", meta.Name())
+	}
+	dt.setRevision(table.revision)
+	table.deleteTrackers, _, _ = table.deleteTrackers.Insert([]byte(trackerName), dt)
+	return nil
+
+}
+
+func (txn *txn) Delete(meta TableMeta, data any) (any, bool, error) {
+	if txn.rootReadTxn == nil {
+		return nil, false, fmt.Errorf("transaction is closed")
+	}
+
+	tableName := meta.Name()
+	revision, err := txn.newRevision(tableName)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Delete from the primary index first to grab the object.
+	// We assume that "data" has only enough defined fields to
+	// compute the primary key.
+	idKey := meta.primaryIndexer().fromObject(object{data: data}).First()
+	idIndexTree := txn.indexWriteTxn(tableName, meta.primaryIndexer().name)
+	obj, existed := idIndexTree.Delete(idKey)
+	if !existed {
+		return nil, false, fmt.Errorf("object not found")
+	}
+
+	// Update revision index.
+	indexTree := txn.indexWriteTxn(tableName, RevisionIndex)
+	if _, ok := indexTree.Delete(revisionKey(obj.revision, idKey)); !ok {
+		panic("BUG: Revision entry not found")
+	}
+
+	// Then update secondary indexes.
+	for index, indexer := range meta.secondaryIndexers() {
+		indexer.fromObject(obj).Foreach(func(key []byte) {
+			if !indexer.unique {
+				key = append(key, idKey...)
+			}
+			txn.indexWriteTxn(tableName, index).Delete(key)
+		})
+	}
+
+	// And finally insert the object into the graveyard.
+	if txn.hasDeleteTrackers(tableName) {
+		graveyardIndex := txn.indexWriteTxn(tableName, GraveyardIndex)
+		obj.revision = revision
+		if old, existed := graveyardIndex.Insert(idKey, obj); existed {
+			txn.indexWriteTxn(tableName, GraveyardRevisionIndex).Delete(revisionKey(old.revision, idKey))
+		}
+		txn.indexWriteTxn(tableName, GraveyardRevisionIndex).Insert(revisionKey(revision, idKey), obj)
+	}
+
+	return obj.data, true, nil
+}
+
+func (txn *txn) Abort() {
+	if txn.writeTxns == nil {
+		return
+	}
+	txn.smus.Unlock()
+	*txn = zeroTxn
+}
+
+func (txn *txn) Commit() {
+	if txn.writeTxns == nil {
+		return
+	}
+
+	// Commit each individual changed index to each table.
+	// We don't notify yet (CommitOnly) as the root needs to be updated
+	// first.
+	for tableIndex, subTxn := range txn.writeTxns {
+		table, ok := txn.tables[tableIndex.table]
+		if !ok {
+			panic("BUG: Table " + tableIndex.table + " not cached")
+		}
+		table.indexes, _, _ =
+			table.indexes.Insert([]byte(tableIndex.index), subTxn.CommitOnly())
+	}
+
+	db := txn.db
+
+	// Acquire the lock on the root tree to sequence the updates to it.
+	db.mu.Lock()
+	rootTxn := db.root.Load().Txn()
+
+	// Insert the modified tables into the root.
+	for name, table := range txn.tables {
+		rootTxn.Insert([]byte(name), *table)
```

Comment:
remark on consistency: I think the following properties hold:

1. the critical section of updating the root is protected from concurrency by `db.mu`
2. Concurrent write transactions can only occur if the set of tables the write transactions cover is disjoint (thanks to the sortable mutexes), and hence there cannot be a conflict of writers here.
3. Read transactions can occur concurrently to a write transaction (even with overlapping tables), but will observe a consistent state (as long as the immutability of objects within the radix is upheld).

Something like this somewhere in the code would help; having the properties formulated makes it easier to think about how they can be broken